platforms/chibios/drivers/wear_leveling/wear_leveling_efl.c

```diff
@@ -0,0 +1,140 @@
+// Copyright 2022 Nick Brassel (@tzarc)
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include <stdbool.h>
+#include <hal.h>
+#include "timer.h"
+#include "wear_leveling.h"
+#include "wear_leveling_internal.h"
+
+static flash_offset_t base_offset = UINT32_MAX;
+
+#if defined(WEAR_LEVELING_EFL_FIRST_SECTOR)
+static flash_sector_t first_sector = WEAR_LEVELING_EFL_FIRST_SECTOR;
+#else  // defined(WEAR_LEVELING_EFL_FIRST_SECTOR)
+static flash_sector_t first_sector = UINT16_MAX;
+#endif // defined(WEAR_LEVELING_EFL_FIRST_SECTOR)
+
+static flash_sector_t sector_count = UINT16_MAX;
+static BaseFlash *    flash;
+
+#ifndef WEAR_LEVELING_EFL_FIRST_SECTOR
+// "Automatic" detection of the flash size -- ideally ChibiOS would have this already, but alas, it doesn't.
+static inline uint32_t detect_flash_size(void) {
+#    if defined(WEAR_LEVELING_EFL_FLASH_SIZE)
+    return WEAR_LEVELING_EFL_FLASH_SIZE;
+#    elif defined(FLASH_BANK_SIZE)
+    return FLASH_BANK_SIZE;
+#    elif defined(FLASH_SIZE)
+    return FLASH_SIZE;
+#    elif defined(FLASHSIZE_BASE)
+#        if defined(QMK_MCU_SERIES_STM32F0XX) || defined(QMK_MCU_SERIES_STM32F1XX) || defined(QMK_MCU_SERIES_STM32F3XX) || defined(QMK_MCU_SERIES_STM32F4XX) || defined(QMK_MCU_SERIES_STM32G4XX) || defined(QMK_MCU_SERIES_STM32L0XX) || defined(QMK_MCU_SERIES_STM32L4XX) || defined(QMK_MCU_SERIES_GD32VF103)
+    return ((*(uint32_t *)FLASHSIZE_BASE) & 0xFFFFU) << 10U; // this register has the flash size in kB, so we convert it to bytes
```

Comment:
If the information at https://mecrisp-stellaris-folkdoc.sourceforge.io/bluepill-diagnostics-v1.6.html#passed-genuine-stm32f103c8t6 is accurate, and genuine STM32F103x8 chips actually report the 64K flash size in this register, then switching to this driver could potentially break some bluepill boards that are currently working with the default `vendor` driver config (although currently the default LD script for F103+stm32duino is for STM32F103x8, it does not reserve the space for EEPROM emulation, therefore it is possible that some firmwares have the size which does not leave enough space for EEPROM if using only the documented 64K). Apparently some people also rely on that feature when actually using the extra space for firmware too: https://github.com/qmk/qmk_firmware/pull/12537#discussion_r616343728.

I cannot verify this on any actual hardware, because the  “STM32F103C8T6” chip that I have reports 128K in the flash size register for some reason (therefore Bluepill Diagnostics decides that it's a fake).

Not really sure what to do about this though. It could be possible to hardcode the expected address at the compile time, so that by default a firmware compiled for STM32F103x8 would always place the EEPROM backing store below the 64K limit (for reliability this requires implementing the flash size check which takes the backing store into account), and a firmware compiled for STM32F103xB would always place the EEPROM backing store just below the 128K limit (instead of trying to check the flash size and potentially placing the backing store below 64K instead of 128K). But then there is not much point in doing this autodetection at all (if the chip ended up smaller than expected, the backing store might not fit; if the chip ended up larger than expected, moving the EEPROM backing store to the end of the real flash does not accomplish much).

BTW, another potential problem with STM32F103 is that the ChibiOS EFL driver for that series apparently does not have proper support for the “XL density” chips (> 512K flash) — those chips have 2 flash banks with separate SR/CR/AR registers, but the EFL driver code in ChibiOS does not have support for that second set of registers.  But I don't have any such chips for testing.